### ISSUE TITLE
Intermittent_activation relic effects break on save/load

### ISSUE THREAD

**Issue by chaosvolt** [cataclysmbnteam/Cataclysm-BN#186]
# Describe the bug

Relics with `intermittent_activation` enchantments take effect normally up until the game is saved, and then loaded. Afterward, the effects will cease to activate.

This appears to be a separate issue from the "intermittent activation relic_data specified directly inside the item breaks saves" bug in that it takes affect even if `enchantment::serialize` is fixed, and can be tested with explicitly-defined enchantments (which avoid the save-breaking bug).

# Steps To Reproduce

1. Use version of branch from https://github.com/cataclysmbnteam/Cataclysm-BN/pull/185, which fixes the save breakages and updates the Cube of Shame to be usable for testing this function.
2. Start game and spawn in a Cube of Shame.
3. Wait 5-10 minutes for healing messages to appear, even at full HP.
4. Save and load game.
5. Resume waiting. No more messages from the spell triggering.

# Expected behavior

Save/load to have no impact on function of intermittent_activation.

# Screenshots

Before saving:
![image](https://user-images.githubusercontent.com/11582235/98037054-a13de500-1de0-11eb-9534-c6df383da583.png)

After saving and loading (in a test build for the PR fixing save breakage):
![image](https://user-images.githubusercontent.com/11582235/98037072-a733c600-1de0-11eb-913c-33bd1bb0bc26.png)

# Versions and configuration

# Additional context

```
- OS: Windows
    - OS Version: 10.0 1803
- Game Version: 7da6104589-dirty [64-bit]
- Graphics Version: Tiles
- Game Language: System language []
- Mods loaded: [
    Dark Days Ahead [dda],
    Disable NPC Needs [no_npc_food]
]
```

Related DDA issue: https://github.com/CleverRaven/Cataclysm-DDA/issues/41261

As mentioned, it seems that fixing the save breakage is easy enough, but it simply exposed an underlying issue, which I was able to eventually confirm affects sources of `intermittent_activation` that would not normally be subject to the save/load bug.

**Comment by Coolthulhu** [comment]
Before save (the json):
```
"intermittent_activation": { "effects": [ { "frequency": "5 minutes", "spell_effects": [ { "id": "AEA_HEAL" } ] } ] }
```
After save (extracted from save file):
```
"intermittent_activation": { "duration": 300, "spell_effects": [ { "id": "AEA_HEAL", "hit_self": false, "once_in": 1, "max_level": -1, "min_level": 0 } ] }
```

Looks like the serialization is just plain incorrect.
I'm not good with Magiclysm syntax, but from the `load` and `serialize` functions themselves, it is obvious that:
* `intermittent_activation` must contain an `effects` array as its only member
* Each member of this array must have a `frequency` field
* `serialize` doesn't wrap them in `effects` nor does it output a `frequency` field

**Comment by chaosvolt** [comment]
That's what I was starting to wonder, the third or fourth time I was testing changes to the code trying to figure out whether I'd actually fixed serialize or not, before I eventually confirmed there were two distinct bugs at work.

I'm not 100% sure if just changing references to `duration` to `frequency` will be enough to fix it, nor do I know whether the save JSON will accept what honestly looks like something that would load-error if inserted into item JSON (which would imply they're interpreted differently at some point), but I can try to tinker with it to see if I can get it to behave sanely.

Though if the first half of point three holds true, that would expect why some early tests I made of renaming `duration` didn't seem to have any effect, I would've only been fixing half the root cause.

Any suggestions for what might be more likely to order the outputted JSON in the desired manner? I'm still not very good at this, sadly.

**Comment by Coolthulhu** [comment]
> Any suggestions for what might be more likely to order the outputted JSON in the desired manner?

Wrap everything you output in an array named `effects`. Then each of the effects in an object.
Think about what would happen if you had more than one intermittent effect. In fact, try adding two such effects to the cube, save and then read the save (CTRL+F "intermittent" to find it). I'd expect the save to error, because you're not wrapping each `fake_spell` into a separate object.

**Comment by chaosvolt** [comment]
Ah, I see. So I'm guessing the rough changes would be:
1. Rename `duration` to `frequency`
2. Re-introduce a reference to `effects` since it was likely actually necessary, and not a mistake like Sim's proposed file change suggested, right before `frequency` is specified
3. Add a call to start an array in between `effects` and ``frequency`
4. Add a call to end an array after the `for` loop that checks for fake spells (uncertain of exact placement for where to put that end_array)

Though, presumably...I'm guessing that start_object and end_object control placement of `{` and `}` just as with array commands and `[`/`]`, but the brackets generated for `AEA_HEAL` are generated automatically by the `for( const fake_spell &sp : pair.second )` section that handles fake spells? Hence why an object starts and ends only once, roughly in the order you'd expect from the function's order, but the save output contains a second pair of brackets correctly-placed for fake spells?

**Comment by chaosvolt** [comment]
Assuming one doesn't need to start and end an object again inside the sequence, rough draft of what it might look like:
```
    if( !intermittent_activation.empty() ) {
        jsout.member( "intermittent_activation" );
        jsout.start_object();
        for( const std::pair<time_duration, std::vector<fake_spell>> pair : intermittent_activation ) {
            jsout.member( "effects" );
            jsout.start_array();
            jsout.member( "frequency", pair.first );
            jsout.member( "spell_effects" );
            jsout.start_array();
            for( const fake_spell &sp : pair.second ) {
                sp.serialize( jsout );
            }
            jsout.end_array();
            jsout.end_array();
        }
        jsout.end_object();
    }
```

This is almost _certainly_ wrong to some degree, but not 100% sure. Part of how wrong it'd be would also depend on:
1. Whether the first `for` section actually does anything sane, given there is generally no reason to have more than one block of `intermittent_activation` inside a single enchantment (or rather, within a single object enclosed inside relic_data that forms something similiar to an enchantment but defined on-the-spot in the relic rather than as an explicit reusable enchantment).
2. Whether or not `start_object` needs to be called a second time in here somewhere.
3. Whether the second `for` section correctly grabs multiple spell `ID`s in a single `effects` section, without creating broken save JSON.

**Comment by Coolthulhu** [comment]
You start the array inside the loop, meaning you're trying to create a new array with the same name for every effect.
You can easily test your hypotheses: compile, start game, generate item, save, find the item in the save. Copy it to a text file so that you can see "in" and "out" side by side.

**Comment by chaosvolt** [comment]
Ah, I see. I'll need to tinker with it in a bit then, still not yet 100% sure how this'll work out but I'll try to test it when I get a branch for it prepared.

**Comment by chaosvolt** [comment]
Cube's version of the JSON:
```
"intermittent_activation": { "effects": [ { "frequency": "5 minutes", "spell_effects": [ { "id": "AEA_HEAL" } ] } ] }
```

Result from save JSON, latest test:
```
"intermittent_activation": { "effects": [ "frequency": 300, "spell_effects": [ { "id": "AEA_HEAL", "hit_self": false, "once_in": 1, "max_level": -1, "min_level": 0 } ] ] }
```

That's actually close, a lil bit at least, looks like if I inject start and end object in the correct places it might finally work.

**Comment by chaosvolt** [comment]
Okay, this is odd...tested this variant of the code:
```
    if( !intermittent_activation.empty() ) {
        jsout.member( "intermittent_activation" );
        jsout.start_object();
        for( const std::pair<time_duration, std::vector<fake_spell>> pair : intermittent_activation ) {
            jsout.member( "effects" );
            jsout.start_array();
            jsout.start_object();
            jsout.member( "frequency", pair.first );
            jsout.member( "spell_effects" );
            jsout.start_array();
            for( const fake_spell &sp : pair.second ) {
                sp.serialize( jsout );
            }
            jsout.end_array();
            jsout.end_object();
            jsout.end_array();
        }
        jsout.end_object();
    }
```

Gave the cube the following relic data:
```
    "relic_data": {
      "passive_effects": [
        { "id": "AEP_EVIL" },
        {
          "has": "HELD",
          "condition": "ALWAYS",
          "mutations": [ "CARNIVORE" ],
          "intermittent_activation": { "effects": [ { "frequency": "5 minutes", "spell_effects": [ { "id": "AEA_HEAL" }, { "id": "AEA_HEAL" } ] } ] }
        },
        {
          "has": "HELD",
          "condition": "ACTIVE",
          "values": [ { "value": "INTELLIGENCE", "add": -4 }, { "value": "PERCEPTION", "add": -4 } ],
          "ench_effects": [ { "effect": "debug_clairvoyance", "intensity": 1 } ]
        }
      ]
    }
```

And this is what the save JSON data returns:
```
"relic_data": { "moves": 100, "charges_per_activation": 1, "passive_effects": [ { "id": "AEP_EVIL" }, { "has": "HELD", "condition": "ACTIVE", "mutations": [  ], "values": [ { "value": "PERCEPTION", "add": -4 }, { "value": "INTELLIGENCE", "add": -4 } ] } ] }
```

Save loads without errors but it's no longer saving the fake spells. Also after loading AEP_EVIL still isn't triggering, so possibly I might need to mess with `load` too.

**Comment by Coolthulhu** [comment]
Think about like this:
* `start_object` is a `{`
* `start_array` is a `[`

If you put a `start_array` in a loop, you're opening as many arrays as there are iterations of the loop.
In
```
"intermittent_activation": { "effects": [ { "frequency": "5 minutes", "spell_effects": [ { "id": "AEA_HEAL" }, { "id": "AEA_HEAL" } ] } ] }
```
What do you have 2 off? `"effects"` or `"spell_effects"`? How many `jsout.member( "effects" );` will happen if you run the array body 2 times?

> But then it won't load, complaining that it wants `frequency` to be saved as a string.

Two solutions:
* (Easier) Add a special case for loading `"frequency"` that is an `int`. 
```if(has_int("frequency"){ load_as_int }else{load_as_string (the current version)}```
* Find out how to save `time_duration` as a string.

**Comment by chaosvolt** [comment]
I see. The two for loops are definitely a bit unusual and probably part of what's messing this up, but I'm not yet sure how to fix those. As for what I had two of, it was two fake spells contained in `spell_effects`, and only one `effects`.

**Comment by Coolthulhu** [comment]
> The two for loops are definitely a bit unusual

Nope.
The problem is that you're creating the `"effects"` array in a loop, inside an object (and not an array). You're creating this:
```
"intermittent_activation": {
  "effects": [ { "frequency": 300, "spell_effects": [ { "id": "AEA_HEAL"] } ],
  "effects": [ { "frequency": 300, "spell_effects": [ { "id": "AEA_HEAL"] } ]
}
```
And this can't exist, because you can't have two members with the same name.

**Comment by chaosvolt** [comment]
Ah, I see. I might have to either figure out how to make them not do that, or else scrap them in favor of aping some other section of the serialize code. The for loops are part of the original code by Korg so I'm not very familiar with it at all.

**Comment by chaosvolt** [comment]
Ah, looking at it further, I'm gonna tinker a bit with moving effects and such before the first for loop, then see if I can add the other proposed changes. If today's further tinkering with it fails, I might give it a break for a bit and see what else I need to tackle to start porting over mods.

**Comment by Coolthulhu** [comment]
Move the `"effects"` object creation (and ending) outside loop. It should be around the loop, not inside it.

**Comment by chaosvolt** [comment]
About to test that, yes. First variant of the test will likely look like this:
```

    if( !intermittent_activation.empty() ) {
        jsout.member( "intermittent_activation" );
        jsout.start_object();
        jsout.member( "effects" );
        jsout.start_array();
        jsout.start_object();
        for( const std::pair<time_duration, std::vector<fake_spell>> pair : intermittent_activation ) {
            jsout.member( "frequency", pair.first );
            jsout.member( "spell_effects" );
            jsout.start_array();
            for( const fake_spell &sp : pair.second ) {
                sp.serialize( jsout );
            }
            jsout.end_array();
        }
        jsout.end_object();
        jsout.end_array();
        jsout.end_object();
    }
```

If that compiles and loads, then I'll next try to work on getting it to load the duration/frequency properly.

**Comment by chaosvolt** [comment]
Tested three variants of the intermittent effect, and grabbed how they were saved for each instance:
```
"intermittent_activation": { "effects": [ { "frequency": "5 minutes", "spell_effects": [ { "id": "AEA_HEAL" } ] } ] }
```
```
"intermittent_activation": { "effects": [ { "frequency": 300, "spell_effects": [ { "id": "AEA_HEAL", "hit_self": false, "once_in": 1, "max_level": -1, "min_level": 0 } ] } ] }
```

```
"intermittent_activation": { "effects": [ { "frequency": "5 minutes", "spell_effects": [ { "id": "AEA_HEAL" }, { "id": "AEA_HEAL" } ] } ] }
```
```
"intermittent_activation": { "effects": [ { "frequency": 300, "spell_effects": [ { "id": "AEA_HEAL", "hit_self": false, "once_in": 1, "max_level": -1, "min_level": 0 }, { "id": "AEA_HEAL", "hit_self": false, "once_in": 1, "max_level": -1, "min_level": 0 } ] } ] }
```

```
"intermittent_activation": { "effects": [ { "frequency": "5 minutes", "spell_effects": [ { "id": "AEA_HEAL" } ] }, { "frequency": "10 minutes", "spell_effects": [ { "id": "AEA_HEAL" } ] } ] }
```
```
"intermittent_activation": { "effects": [ { "frequency": 300, "spell_effects": [ { "id": "AEA_HEAL", "hit_self": false, "once_in": 1, "max_level": -1, "min_level": 0 }, { "id": "AEA_HEAL", "hit_self": false, "once_in": 1, "max_level": -1, "min_level": 0 } ] } ] }
```

The last one is an odd one out, in that it rejects there being two entries in the `effects` block and lumps them together as one effect with two `spell_effects` instead.
